### PR TITLE
Backporting changes of CRM-17836 in 4.6

### PR DESCRIPTION
**Pull request by monishdeb**
---
- CRM-17836: custom fields attached to relationships are unsearchable
  https://issues.civicrm.org/jira/browse/CRM-17836

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
looks like a minor enotice is wreaking havoc

**Comment by monishdeb** [comment]
Done with test fix

**Comment by eileenmcnaughton** [comment]
This fixes the described problem. However, I'm getting some regression on date fields - doing a bit more digging...

**Comment by eileenmcnaughton** [comment]
OK - I wasn't getting searches working on custom date fields with this patch. We had our installs patched for CRM-17116 so I don't know what core was like before your patch but with 4.6 + this I was seeing misinterpretation of custom date fields for both range & not range fields.

I  got both your patch and custom date relative searching working together - I didn't bother to try to get it working without the CRM-17116 improvement (as I we were wanting CRM-17116 for our customers.). My working patch is at https://github.com/civicrm/civicrm-core/compare/4.6...eileenmcnaughton:CRM-17116-46-plus-monish-patch?expand=1

I am waiting to merge in some of the 'noise' (syncing with master) before I create a PR for your patch & mine & then hopefully you can do some QA & see how it is.

**Comment by monishdeb** [comment]
@eileenmcnaughton thanks a lot for adding test-coverage in your patch for date-range which I have extended to cover all other custom field of data-type which support search by range. BUT the patch was breaking for fields other than data-type, that's why I have retained my patch and extended the CRM test to cover all scenario, also did some few fixes to your test-case. Please have a look.

**Comment by eileenmcnaughton** [comment]
I'll give it another spin later

**Comment by eileenmcnaughton** [comment]
NB the code I used was already in 4.7 &  the test should have been too?

**Comment by eileenmcnaughton** [comment]
As an aside I tend to be pretty aggressive about doing 

git fetch -f oriign
git rebase -i origin/4.6

on my PRs & then squashing commits together from that screen - which gets rid of the 'minor fix' ones.

You have to do git push -f but I think that's OK in that context

File: CRM/Contact/BAO/Query.php [hunk 1]
```diff
@@ -1489,6 +1489,15 @@ public static function convertFormValues(&$formValues, $wildcard = 0, $useEquals
...
       }
+      elseif (substr($id, 0, 7) == 'custom_'
+        &&  (
+          substr($id, -9, 9) == '_relative'
+          || substr($id, -5, 5) == '_from'
+          || substr($id, -3, 3) == '_to'
+        )
+      ) {
+        self::convertCustomRelativeFields($formValues, $params, $values, $id);
+      }
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: The missing piece to support non-range custom date fields is to add

      elseif (!empty($values) && !is_array($values) && self::isCustomDateField($id) ) {
        $params[] = array($id, '=', CRM_Utils_Date::processDate($values), 0, $wildcard);
      }

In 4.7 the date is already formatted to a consistent date format by the time it hits this point by the datepicker replacement.

Comment (monishdeb) [inline CRM/Contact/BAO/Query.php]: Does it happen in 4.6 too ?

Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: nope - but those lines above will format it in 4.6 - try searching on a custom date field that is not set up to search by range in 4.6 with your patch - the range custom fields are fine - but the ones that aren't do odd things with the date

File: CRM/Contact/BAO/Query.php [hunk 2]
```diff
@@ -1489,6 +1489,18 @@ public static function convertFormValues(&$formValues, $wildcard = 0, $useEquals
...
+      elseif (!empty($values) && !is_array($values) && self::isCustomDateField($id) ) {
+        $params[] = array($id, '=', CRM_Utils_Date::processDate($values), 0, $wildcard);
+      }
```
Comment (monishdeb) [inline CRM/Contact/BAO/Query.php]: Done :)

**Comment by eileenmcnaughton** [comment]
Except for an annoying style error :-)

FWIW if you do rebase -i you can squash that 'optimizaion' commit into the main commit

**Comment by eileenmcnaughton** [comment]
I put your commit through as a new PR in the hope we can finally get it merged 

https://github.com/civicrm/civicrm-core/pull/7693

**Comment by monishdeb** [comment]
Oops I already squashed those commit :/ .. sorry for delay

**Comment by eileenmcnaughton** [comment]
no worries - I thought you would have gone to bed!

**Comment by monishdeb** [comment]
Closing in favour of #7693

**Pull request closed without merging**